packages/trie/src/trie.ts

```diff
@@ -883,8 +883,28 @@ export class Trie {
   /**
    * Saves the nodes from a proof into the trie.
    * @param proof
+   * @deprecated Use `updateTrieFromProof`
    */
   async fromProof(proof: Proof): Promise<void> {
+    await this.updateTrieFromProof(proof, false)
+
+    if (equalsBytes(this.root(), this.EMPTY_TRIE_ROOT) && proof[0] !== undefined) {
+      let rootKey = Uint8Array.from(this.hash(proof[0]))
+      // TODO: what if we have keyPrefix and we set root? This should not work, right? (all trie nodes are non-reachable)
+      rootKey = this._opts.keyPrefix ? concatBytes(this._opts.keyPrefix, rootKey) : rootKey
+      this.root(rootKey)
+      await this.persistRoot()
+    }
+    return
+  }
+
+  /**
+   * Updates a trie from a proof
+   * @param proof The proof
+   * @param safe If `true`, verifies that the root key of the proof matches the trie root. Throws if this is not the case.
+   * @returns The root of the proof
+   */
+  async updateTrieFromProof(proof: Proof, safe: boolean = false) {
```

Comment:
I don't think this works, because that might alter the state root of the trie. You have to input raw nodes here. If you would put key/values into a partial trie, I think it either throws (but thats an edge case) or it alters the state root (and the latter is likely not what we want). (Have to verify this). For practical purposes it might work though to put the key/values directly, but I'm not sure if you want to change the root there as well.

(I have to verify these claims though :smile: :+1: )